### ISSUE TITLE
Is it safe to derive from Plot and use in FormsPlot.Reset()?

### ISSUE THREAD

**Issue by nightroman** [ScottPlot/ScottPlot#1983]
**ScottPlot Version:** 4.1.52

**Operating System:** Windows 10

**Application Type:** console based Far Manager + FarNet API for .NET modules and scripting

**Question:** Is it safe to derive from Plot and use in FormsPlot.Reset()?

[FarNet.ScottPlot]: https://github.com/nightroman/FarNet.ScottPlot

I work on [FarNet.ScottPlot] making ScottPlot somewhat friendlier for scripts in PowerShell, JavaScript, F#.
So far the most convenient way is having my type `FormPlot` derived from `Plot` and exposing it to scripts.
It addition to `Plot` base members, it adds new members. It works well. But I have some concerns.

Why it is convenient. In most cases scripts have to load just the derived type assembly.
Then all ScottPlot plot properties and methods are available dynamically via an instance.
(more relevant to PowerShell and JavaScript, F# is compiled and needs all references)

- [JavaScript](https://github.com/nightroman/FarNet.ScottPlot/blob/01aa16e73b39e70abfc7a71fdaa581be50826198/samples-JavaScript/Scatter.js#L3)
- [PowerShell](https://github.com/nightroman/FarNet.ScottPlot/blob/01aa16e73b39e70abfc7a71fdaa581be50826198/samples-PowerShell/Scatter.far.ps1#L7)

Why concerns:

(1) The type `Plot` is not sealed but it does not have any virtual members.
So it "looks" like not sealed "accidentally" and may become sealed one day.

Q1: Can I count on `Plot` being not sealed?

(2) The only way to inject my plot instance into `FormsPlot` control is `Reset(Plot)`.
This works. But what if I hijack a feature designed for something else?
Otherwise I'd expect a constructor like `FormsPlot(Plot)` provided.

Q2: Is it fine to inject my plot in a new `FormsPlot` using `Reset(Plot)`?

- [source](https://github.com/nightroman/FarNet.ScottPlot/blob/01aa16e73b39e70abfc7a71fdaa581be50826198/src/FormPlot.cs#L60)

**Comment by bclehmann** [comment]
I think your use-case is neither intended nor supported, but also unlikely to break. Perhaps `Plot` should be `sealed`, but I don't anticipate it will become `sealed` for a few reasons:

- It would be a breaking change
- It wouldn't provide much benefit
- Most substantial changes are probably going to be for ScottPlot 5, not 4.1.x

As for your use of `Reset(Plot)`, you're correct that that method isn't intended to allow swapping out functionality, as `Plot` is not intended to be used polymorphically. It's instead to allow loading in an existing plot. This could be used to let a user click a small plot and blow it up to a larger window, for example. `FormsPlot(Plot)` would be good idea, although it's not very easy to call the constructors of  `WinForms` controls, nor is it easy for the other platforms that ScottPlot supports. Even so, switching the plot object on the fly may itself be useful. Perhaps a better method name is `LoadFrom(Plot plot)`.

So, in summary, you're doing something that (unless @swharden disagrees) ScottPlot is not intended to support through means that ScottPlot is not intended to allow. I don't think `Plot` will become `sealed`, nor will it be substantially changed until ScottPlot 5, so I think you should be fine. Especially as it looks like your `FormsPlot` doesn't seem to depend on non-public members.

That said, given that it looks like your `FormsPlot` doesn't seem to depend on `Plot` internals, you could probably achieve your goals without inheriting from `Plot`, instead creating a wrapper class.

**Comment by nightroman** [comment]
@bclehmann 

> Please note that me saying you should be fine is a personal estimation of how likely it is that things will break, not assurance that they won't or a promise of support if they do. You'll have to weigh the risks yourself.

Thank you for the detailed answer. It's assuring to some extent. Hopefully you'll keep in mind that your current design, intentional or not, has useful applications. But things change, I understand the risks.

> ... you could probably achieve your goals without inheriting from Plot, instead creating a wrapper class.

That would be the safer way. Yet `Plot` has too many members to wrap. So for now I'll go with inheritance and will wrap later if there is no other way.

Thank you for the great library!

**Comment by swharden** [comment]
Hi @nightroman 👋 sorry for the delay in my response

Thanks @bclehmann for your extensive answers. I think you've already said all the good stuff, and I'll just chime in to give some "official" input.

> Q1: Can I count on Plot being not sealed?

Yes! I have no plans to seal it 🚀 

If you want to submit a PR adding a test to confirm/ensure it is not sealed, you're welcome to

https://www.geeksforgeeks.org/c-sharp-program-to-check-a-specified-class-is-a-sealed-class-or-not/

> what if I hijack a feature designed for something else?

Similarly, ScottPlot4 is stable and I have no plans to change it, so you can rely on it! 🚀 

That response contains a lot of rocket ships, so this is good news! Your [FarNet.ScottPlot](https://github.com/nightroman/FarNet.ScottPlot) looks awesome - it's always cool to see how other people use this library.

**Comment by nightroman** [comment]
@swharden Thank you for the great tool and the encouraging answer!

> If you want to submit a PR adding a test to confirm/ensure it is not sealed, you're welcome to

I will certainly do!

Having played more with scripting ScottPlot using [FarNet.ScottPlot](https://github.com/nightroman/FarNet.ScottPlot) I'm still sure that the "unusual" design decision of inheriting is best for scripts. Definitely for PowerShell and JavaScript at least.
